### PR TITLE
Refactor import of BPMN model

### PR DESCRIPTION
**Pull request by rds-greg**
1. Handle resources safely.
2. Handle exceptions appropriately, exit early as required.
3. Use generics where possible.

### CODE REVIEW

File: modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java [hunk 1]
```diff
@@ -150,121 +158,240 @@ public ResultListDataRepresentation getModelsToIncludeInAppDefinition() {
...
+      {
+        String formkey = eeForm.getAttributeValue(null, RdsForms.FORM_KEY);
+        List existingFormModels = modelRepository.findByKeyAndType(formkey, AbstractModel.MODEL_TYPE_FORM_RDS);
+        if (CollectionUtils.isNotEmpty(existingFormModels))
+        {
+          existingForms.add(formkey);
+        }
+      }
+
+      // Return validation result if process with same key or any form with same form key exists if overwrite is false
```
Comment (rds-greg) [inline modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java]: This comment looks out of date :/

File: modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java [hunk 2]
```diff
@@ -150,121 +158,240 @@ public ResultListDataRepresentation getModelsToIncludeInAppDefinition() {
...
     }
+  }
 
-    protected ModelRepresentation createModelRepresentation(AbstractModel model) {
-        ModelRepresentation representation = null;
-        if (model.getModelType() != null && model.getModelType() == 3) {
+  protected ModelRepresentation createModelRepresentation(AbstractModel model)
+  {
+    ModelRepresentation representation;
+    if (model.getModelType() != null && model.getModelType() == AbstractModel.MODEL_TYPE_APP)
```
Comment (rds-greg) [inline modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java:392]: Remove magic number, just want to check if this is the correct constant...

File: modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java [hunk 3]
```diff
@@ -150,121 +158,240 @@ public ResultListDataRepresentation getModelsToIncludeInAppDefinition() {
...
+  }
+
+  private Map<String, ModelInfo> validateAndSaveForms(Process process, boolean overwrite)
+      throws BpmnFileValidationException
+  {
+    Map<String, ModelInfo> formKeyMap = new HashMap<>();
+    List<String> existingForms = new ArrayList<>();
+    List<ExtensionElement> forms = process.getExtensionElements().get(RdsForms.FORM_ELEMENT);
+
+    if(CollectionUtils.isNotEmpty(forms))
```
Comment (rds-greg) [inline modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java:263]: Handles what was a null pointer exception where there were no forms in the bpmn

File: modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java [hunk 4]
```diff
@@ -150,121 +158,240 @@ public ResultListDataRepresentation getModelsToIncludeInAppDefinition() {
...
+        if (xtr != null)
+        {
+          xtr.close();
+        }
+      }
+      catch (XMLStreamException e)
+      {
+        throw new BpmnFileException("Unable to close XMLStreamReader", e);
+      }
+    }
```
Comment (ruiboy) [inline modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java:231]: +1 for the dilligent stream handling

**Review by simon998yang — COMMENTED** [review]
Just be aware when updating flowable files (our newly created files are fine). We didnot reformat/refactor extensively flowable files as this eventually would cause merge conflict when we need to upgrade flowable.

**Review by simon998yang — APPROVED** [review]

File: modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java [hunk 5]
```diff
@@ -150,121 +158,240 @@ public ResultListDataRepresentation getModelsToIncludeInAppDefinition() {
...
+        String formkey = eeForm.getAttributeValue(null, RdsForms.FORM_KEY);
+        List existingFormModels = modelRepository.findByKeyAndType(formkey, AbstractModel.MODEL_TYPE_FORM_RDS);
+        if (CollectionUtils.isNotEmpty(existingFormModels))
+        {
+          existingForms.add(formkey);
+        }
+      }
+
+      // Return validation result if process with same key or any form with same form key exists if overwrite is false
+      if (!overwrite && CollectionUtils.isNotEmpty(existingForms))
```
Comment (ruiboy) [inline modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java]: why no longer failing if process key already exists, ie validationResult.size() > 0  ?

Comment (ruiboy) [reply modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java]: scrap that ... i see you are failing earlier in that case.   (so much refactoring - got lost ;))

File: modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java (hunk 2 above)
Comment (ruiboy) [reply modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java:392]: looks good to me

File: modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java [hunk 6]
```diff
@@ -150,121 +158,240 @@ public ResultListDataRepresentation getModelsToIncludeInAppDefinition() {
...
-    public Object importProcessModel(HttpServletRequest request, MultipartFile file, boolean overwrite) {
+  public Object importProcessModel(HttpServletRequest request, MultipartFile file, boolean overwrite)
+  {
+    try
+    {
+      BpmnModel bpmnModel = readFromFile(file);
+
+      // Its got information about how to display the bpmn diagram in the modeler (x, y coordinates, height and width etc.)
+      // so call
+      if (MapUtils.isNotEmpty(bpmnModel.getLocationMap()))
```
Comment (ruiboy) [inline modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java]: you've inverted the logic here.  was that your intention?

File: modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java (hunk 1 above)
Comment (ruiboy) [reply modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java]: does now as you handle the duplicate process elsewhere

Comment (rds-greg) [reply modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java]: I'll update the comment to reflect the changes

File: modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java (hunk 6 above)
Comment (rds-greg) [reply modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java]: Oops, no. I'll fix that up

**Review by ruiboy — APPROVED** [review]

**Pull request merged**